http-api/src/patches.rs

```diff
@@ -52,9 +156,162 @@ async fn patches_handler(ctx: Context, urn: Urn) -> Result<impl Reply, Rejection
+    storage: &Storage,
+    include_changeset: bool,
+) -> patch::Patch<Option<Changeset>, project::PeerInfo> {
+    let revisions = patch
+        .revisions
+        .into_iter()
+        .map(|revision| {
+            // Resolves Merge structs with PeerInfo
+            let merges = revision
+                .merges
+                .iter()
+                .map(|merge| {
+                    let peer = project::PeerInfo::get(&merge.peer, meta, storage);
+                    patch::Merge {
+                        peer,
+                        commit: merge.commit,
+                        timestamp: merge.timestamp,
+                    }
+                })
+                .collect();
+
+            // Locates the browser at the Oid of the current revision.
+            if let Err(e) = browser.rev(git::Rev::Oid(*revision.oid)) {
+                tracing::warn!(
+                    "Failed to set the browser history at {}: {}",
+                    *revision.oid,
+                    e
+                );
+
+                return patch::Revision {
+                    id: revision.id,
+                    peer: revision.peer,
+                    oid: revision.oid,
+                    base: revision.base,
+                    comment: revision.comment,
+                    discussion: revision.discussion,
+                    reviews: revision.reviews,
+                    timestamp: revision.timestamp,
+                    changeset: None,
+                    merges,
+                };
+            }
+
+            let mut changeset: Option<Changeset> = None;
+
+            // This feature flag, allows us to only generate diffs for e.g. single patch retrieval and skip all this for patch listing.
+            if include_changeset {
+                // Iterates over commits headers and retrieves each commit details until it gets to the head of the default branch
+                // If radicle_source::commit returns a None the commit won't be collected.
+                // TODO: Probably we can handle a possible error better, but for now we handle it in the interface
+                let commits = if let Ok(commits) = radicle_source::commits::<PeerId>(browser, None)
+                {
+                    Some(
+                        commits
+                            .headers
+                            .iter()
+                            .take_while(|header| header.sha1 != *revision.base)
+                            .filter_map(|header| radicle_source::commit(browser, header.sha1).ok())
+                            .collect::<Vec<Commit>>(),
+                    )
+                } else {
+                    None
+                };
+
+                // Gets the entire diff between the default branch head and the revision Oid.
+                // TODO: Probably we can handle a possible error better, but for now we handle it in the interface
+                let diff = browser.diff(*revision.base, *revision.oid).ok();
```

Comment:
There's no need to do a diff if `commits == None`, since the changeset will be `None`! Sorry didn't notice this before :)

Reply:
I would probably simplify `Changeset::new` and just have it accept the types without the `Option` as well, otherwise we're allowing stuff like this.

Reply:
Nice, good catch, was able to reduce nesting by removing if branches and make the code clearer in 36522db5b0a90a8216ecb3f222f8b2a3a7fac9c8